optimade/filterparser/lark_parser.py

```diff
@@ -1,33 +1,42 @@
-import os
-import re
-from glob import glob
-
+from pathlib import Path
 from lark import Lark, Tree
-
-parser = {}
-for name in glob(os.path.join(os.path.dirname(__file__), "../grammar", "*.g")):
-    with open(name) as f:
-        ver = tuple(
-            int(n)
-            for n in re.findall(r"\d+", str(os.path.basename(name).split(".g")[0]))
-        )
-        parser[ver] = Lark(f.read())
+from collections import defaultdict
 
 
 class ParserError(Exception):
     pass
 
 
+def get_versions():
+    dct = defaultdict(dict)
+    for filename in Path(__file__).parent.joinpath('../grammar').glob('*.lark'):
+        tags = filename.stem.lstrip('v').split('.')
+        version = tuple(map(int, tags[:3]))
+        variant = 'default' if len(tags) == 3 else tags[-1]
+        dct[version][variant] = filename
+    return dct
+
+
+available_parsers = get_versions()
+
+
 class LarkParser:
-    def __init__(self, version=None):
-        if version is None:
-            self.version = sorted(parser.keys())[-1]
-            self.lark = parser[self.version]
-        elif version in parser:
-            self.lark = parser[version]
-            self.version = version
-        else:
+    def __init__(self, version=None, variant='default'):
+
+        version = version if version else max(available_parsers.keys())
+
+        if version not in available_parsers:
             raise ParserError(f"Unknown parser grammar version: {version}")
+
+        if variant not in available_parsers[version]:
+            raise ParserError(f"Unknown variant of the parser: {variant}")
+
+        self.version = version
+        self.variant = variant
+
+        with open(available_parsers[version][variant]) as file:
+            self.lark = Lark(file)
```

Comment:
In the older version here, it seemed `file.read()` was called as input for `Lark()`, but this is not actually necessary, or?

Reply:
Previously, when you imported the module, all versions of the grammar had been loaded and the whole `Lark` object has been stored in the `parser` dictionary. Now is stores only the available versions and creates a Parser for the specific versions when you call the constructor of `LarkParser`.
In this way maybe we can gain some performance (by not loading all the possible grammars when we want to use only a specific one)...
To be honest I changed it because a tiny mistake braked the whole module when I worked on the grammar.

Reply:
Sure - it was only an observational comparison :) If this works, then great! It saves us some extra characters.